core/src/types/writer.rs

```diff
@@ -90,14 +91,22 @@ impl Writer {
         Ok(Writer { inner: w })
     }
 
-    /// Write into inner writer.
-    pub async fn write(&mut self, bs: impl Into<Bytes>) -> Result<()> {
-        let mut bs = bs.into();
+    pub async fn write(&mut self, bs: Buffer) -> Result<()> {
+        let mut bs = bs;
         while !bs.is_empty() {
-            let n = self.inner.write(bs.clone().into()).await?;
+            let n = self.inner.write_dyn(bs.clone()).await?;
             bs.advance(n);
         }
+        Ok(())
+    }
 
+    pub async fn write_from(&mut self, bs: &mut impl Buf) -> Result<()> {
+        while bs.has_remaining() {
+            let chunk_ref: &[u8] = bs.chunk();
+            let static_ref: &'static [u8] = unsafe { mem::transmute(chunk_ref) };
```

Comment:
Should we use unsafe to avoid an extra copy?

Reply:
No, it's better not to do this.